File: 20230701_Saturday/20230701.rb
# Leetcode: 2305. Fair Distribution of Cookies.
# https://leetcode.com/problems/fair-distribution-of-cookies/
# = = = = = = = = = = = = = =
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# @param {Integer[]} cookies
# @param {Integer} k
# @return {Integer}
# TLE.
def distribute_cookies(cookies, k)
  @dp = {}
  @key = Array.new(k+2,0) # index, max
  @c = cookies
  @k = k
  @min = Float::INFINITY
  dfs()
  return @min
end


def dfs
  if @key[0] == @c.size
    @min = @key[1] if @key[1] < @min
    return
  end
  return @dp[@key] if @dp[@key]
  (0...@k).each do |j|
    j = j + 2 # Offset from i and max;

    # Put element in j pocket.
    
    # 1, 2, 3; i, pocket, max;
    i = @key[0] # 1) Save i.
    old_val = @key[j] # 2) Save pocket.
    old_max = @key[1] # 3) Save max.

    # 1, 2, 3; i, pocket, max;
    @key[0] = i + 1 # 1) Change i;
    @key[j] += @c[i] # 2) Change poket.
    @key[1] = @key[j] if @key[j] > old_max # 3) Chage pocket.

    dfs()

    # 1, 2, 3; i, pocket, max;
    @key[0] = i # 1) Restore i.
    @key[j] = old_val # 2) Restore pocket.
    @key[1] = old_max # 3) Restore max.
  end
  @dp[@key.clone] = true
end

File: 20230701_Saturday/20230701.py
# Leetcode: 2305. Fair Distribution of Cookies.
# https://leetcode.com/problems/fair-distribution-of-cookies/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 1594 ms, faster than 36.20% of Python3 online submissions for Fair
# Distribution of Cookies.
# Memory Usage: 16.5 MB, less than 10.82% of Python3 online submissions for Fair
# Distribution of Cookies.
# 2023.07.01 Daily Challenge.
class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        # Copied from:
        # https://leetcode.com/problems/fair-distribution-of-cookies/solution/
        cur = [0] * k
        n = len(cookies)

        def dfs(i, zero_count):
            # If there are not enough cookies remaining, return `float('inf')` 
            # as it leads to an invalid distribution.
            if n - i < zero_count:
                return float('inf')
            
            # After distributing all cookies, return the unfairness of this
            # distribution.
            if i == n:
                return max(cur)
            
            # Try to distribute the i-th cookie to each child, and update answer
            # as the minimum unfairness in these distributions.
            answer = float('inf')
            for j in range(k):
                zero_count -= int(cur[j] == 0)
                cur[j] += cookies[i]
                
                # Recursively distribute the next cookie.
                answer = min(answer, dfs(i + 1, zero_count))
                
                cur[j] -= cookies[i]
                zero_count += int(cur[j] == 0)
            
            return answer
        
        return dfs(0, k)
